[WGSL] Minor improvements to MetalFunctionWriter

tadeuzagallo opened the pull request:
#### d2ccfc205f0c7d006240466c1321ccc0a063fd45
<pre>
[WGSL] Minor improvements to MetalFunctionWriter
<a href="https://bugs.webkit.org/show_bug.cgi?id=251365">https://bugs.webkit.org/show_bug.cgi?id=251365</a>
rdar://104821399

Reviewed by Myles C. Maxfield.

This PR combines a few tiny changes to MetalFunctionWriter as we work
towards supporting compute shaders. Namely:
1. Serialize `void` return types
2. Serialize the `global_invocation_id` builtin
3. Serialize `uint32` literals
4. Only use the buffer index calculation function for vertex shaders

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::visitArgumentBufferParameter):

Canonical link: <a href="https://commits.webkit.org/259622@main">https://commits.webkit.org/259622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/081081ad7ad925f7605ab54b4585b86c08c81a66

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105434 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38306 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114689 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5441 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97740 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111192 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12129 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95107 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39611 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93995 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26742 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7813 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28098 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7935 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13954 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47646 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6638 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9727 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -71,6 +71,7 @@ class FunctionDefinitionWriter : public AST::Visitor {
     void visit(AST::Float32Literal&) override;
     void visit(AST::IdentifierExpression&) override;
     void visit(AST::Int32Literal&) override;
+    void visit(AST::Uint32Literal&) override;
```
🆒

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -95,6 +96,7 @@ class FunctionDefinitionWriter : public AST::Visitor {
     AST::ShaderModule& m_shaderModule;
     Indentation<4> m_indent { 0 };
     std::optional<AST::StructRole> m_structRole;
+    std::optional<AST::StageAttribute::Stage> m_entryPointStage;
```
🆗

tadeuzagallo commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -253,7 +269,14 @@ void FunctionDefinitionWriter::visit(AST::TypeDecl& type)
+        m_stringBuilder.append("*");
+        return;
+    }
+
```
I'll remove this bit as I implemented it correctly in #9337

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -146,8 +155,6 @@ void FunctionDefinitionWriter::visit(AST::StructDecl& structDecl)
         IndentationScope scope(m_indent);
         for (auto& member : structDecl.members()) {
             m_stringBuilder.append(m_indent);
-            if (m_structRole == AST::StructRole::ArgumentBuffer)
```
Are we sure we want ArgumentBuffer to be a member of Role? Doesn't it kind of leak MSL concepts into the AST? Shouldn't we be saying "BindGroup" instead?

Maybe leaking MSL concepts into the AST is a lost cause? 😕

tadeuzagallo replied:
I think this is also a side effect of the fact that I'm still learning how to translate those terms. I'm also not 100% we can avoid the leak, but it doesn't seem strictly necessary for now.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -193,11 +200,17 @@ void FunctionDefinitionWriter::visit(AST::BuiltinAttribute& builtin)
         return;
     }
 
+    if (builtin.name() == "global_invocation_id"_s) {
+        m_stringBuilder.append("[[thread_position_in_grid]]");
+        return;
+    }
```
🆗

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -206,15 +219,18 @@ void FunctionDefinitionWriter::visit(AST::StageAttribute& stage)
         break;
     case AST::StageAttribute::Stage::Compute:
-        m_stringBuilder.append("[[compute]]");
+        m_stringBuilder.append("[[kernel]]");
```
👍

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -253,7 +269,14 @@ void FunctionDefinitionWriter::visit(AST::TypeDecl& type)
+        visit(*type.maybeElementType());
+        m_stringBuilder.append("*");
+        return;
+    }
```
Isn't this wrong, as per our discussion this morning?

tadeuzagallo replied:
Yes, I left a comment above, I fixed this in a separate PR and will remove from this one before merging.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -336,6 +359,7 @@ void FunctionDefinitionWriter::visit(AST::StructType& structType)
 
 void FunctionDefinitionWriter::visit(AST::ReferenceType& type)
 {
+    m_stringBuilder.append("device ");
```
I think this is wrong.

(I'm assuming "Reference type"s correspond to the concept of the same name in the spec: https://gpuweb.github.io/gpuweb/wgsl/#ref-ptr-types

If so, then a reference can refer to an lvalue in any address space, not just the `device` address space.

The reference type itself probably needs to remember what address space it's point into to emit the correct MSL.

tadeuzagallo replied:
That's fair enough, I'll add a FIXME.

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -352,7 +376,7 @@ void FunctionDefinitionWriter::visit(AST::Parameter& parameter)
 
 void FunctionDefinitionWriter::visitArgumentBufferParameter(AST::Parameter& parameter)
 {
-    m_stringBuilder.append("const device ");
+    m_stringBuilder.append("constant ");
```
👍

litherum commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -454,6 +478,12 @@ void FunctionDefinitionWriter::visit(AST::Int32Literal& literal)
     m_stringBuilder.append(literal.value());
 }
 
+void FunctionDefinitionWriter::visit(AST::Uint32Literal& literal)
+{
+    // FIXME: this might not serialize all values correctly
+    m_stringBuilder.append(literal.value());
+}
```
Sure

tadeuzagallo committed:
[WGSL] Minor improvements to MetalFunctionWriter
https://bugs.webkit.org/show_bug.cgi?id=251365
rdar://104821399

Reviewed by Myles C. Maxfield.

This PR combines a few tiny changes to MetalFunctionWriter as we work
towards supporting compute shaders. Namely:
1. Serialize `void` return types
2. Serialize the `global_invocation_id` builtin
3. Serialize `uint32` literals
4. Only use the buffer index calculation function for vertex shaders

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::visitArgumentBufferParameter):

Canonical link: https://commits.webkit.org/259622@main
